controller.py

```diff
@@ -273,18 +276,52 @@ def library_details(self, uuid, library=None):
         )
         return dict(uuid=uuid, basic_info=basic_info, urls_and_contact=urls_and_contact, stages=stages)
 
-    def _contact_email(self, id):
-        contact_email = None
-        query = self._db.query(Resource).join(
-                Hyperlink, Resource.id==Hyperlink.resource_id
-            ).filter(
-                Hyperlink.library_id==id, Hyperlink.rel == Hyperlink.INTEGRATION_CONTACT_REL
+    def _contact_email(self, resource):
+        if resource and resource.href:
+            return resource.href.split("mailto:")[1]
+
+    def _validated_at(self, resource):
+        validated_at = "Not validated"
+        validation = get_one(self._db, Validation, resource=resource)
+        if validation:
+            validated_at = validation.started_at
+        return validated_at
+
+    def _get_resource(self, library_id, rel):
+        hyperlink = self._get_hyperlink(library_id, rel)
+        if hyperlink:
+            query = self._db.query(Resource).filter(
+                Resource.id==hyperlink.resource_id
             )
+            return query.first()
 
-        if query.first():
-            contact_email = query.first().href.split("mailto:")[1]
+    def _get_hyperlink(self, library_id, rel):
+        query = self._db.query(Hyperlink).filter(
+            Hyperlink.library_id==library_id,
+            Hyperlink.rel==rel
+        )
+        hyperlink = query.first()
+        return hyperlink
```

Comment:
This code is fine but it might be better to make it a method of the Library class, `Library.hyperlink(rel)`. That way other bits of code can use it.